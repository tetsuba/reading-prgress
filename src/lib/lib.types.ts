import { ApiTypes } from '../api/api-types'

type LibTypes = {
    value: string
}

type TargetValueTypes = Record<
    keyof Pick<ApiTypes, 'firstName' | 'lastName' | 'email' | 'password'>,
    Pick<LibTypes, 'value'>
>

export type RegistrationFormTypes = EventTarget & TargetValueTypes
