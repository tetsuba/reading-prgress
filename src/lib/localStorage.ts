export const AUTH_KEY = 'reading-progress-auth-key'
export default {
    save: function (value: string): void {
        localStorage.setItem(AUTH_KEY, value)
    },
    get: function (): string {
        return localStorage.getItem(AUTH_KEY) || ''
    },
    remove: function (): void {
        localStorage.removeItem(AUTH_KEY)
    }
}
