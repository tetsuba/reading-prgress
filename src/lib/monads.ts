type Func<V, R> = (val: V) => R

type Task<T> = {
    chain: <C>(f: Func<T, C>) => C
    map: <M>(f: Func<T, M>) => Task<M>
    fold: <F>(f: Func<T, F>) => F
    emit: () => T
    concat: <C>(a: C[]) => Task<Array<C>>
}

type Left<T> = {
    chain: <C>(f: Func<T, C>) => Left<T>
    map: <M>(f: Func<T, M>) => Left<T>
    fold: <R>(l: (v: T) => R, r: (v: T) => R) => R
}

type Right<T> = {
    chain: <C>(f: Func<T, C>) => C
    map: <M>(f: Func<T, M>) => Right<M>
    fold: <R>(l: (v: T) => R, r: (v: T) => R) => R
}

type MonadTask = <T>(data: T) => Task<T>
type MonadRight = <T>(data: T) => Right<T>
type MonadLeft = <T>(data: T) => Left<T>

export const Task: MonadTask = (val) => ({
    /**
     * Returns the Box for chaining.
     */
    val,
    chain: (f) => f(val),
    map: (f) => Task(f(val)),
    fold: (f) => f(val),
    emit: () => val,
    concat: (a) => Task(a.concat(val as [])),
    debug: () => {
        console.log(val)
        return Task(val)
    }
})

export const Right: MonadRight = (val) => ({
    chain: (f) => f(val),
    map: (f) => Right(f(val)),
    fold: (l, r) => r(val)
})

export const Left: MonadLeft = (val) => ({
    chain: (v) => Left(val),
    map: (v) => Left(val),
    fold: (l, r) => l(val)
})

export function isUndefined<T>(data: T | undefined) {
    if (data === undefined) {
        console.error('Args is undefined')
        return Left(data as Exclude<T, undefined>)
    } else {
        return Right(data as Exclude<T, undefined>)
    }
}
