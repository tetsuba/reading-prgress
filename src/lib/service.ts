import axios from 'axios'
import store from '../store/store'
import { updateUser } from '../store/user/userSlice'
import ls from './localStorage'
import { HistoryTypes } from '../store/book/bookSlice'
let origin
/* c8 ignore next 6 */
if (import.meta.env.DEV) {
    origin =
        location.port === '5173' ? 'http://localhost:3001' : location.origin
} else {
    origin = 'http://18.132.193.58:3001'
}

const BASE_URL = `${origin}/api/reading/`

const URL_REGISTER = BASE_URL + 'user/register?'
const URL_LOGIN = BASE_URL + 'user/login'
const URL_USER = BASE_URL + 'user'
const URL_BOOK = BASE_URL + 'book'
const URL_SIGHT_WORDS = BASE_URL + 'sightWords'
const URL_TRACKER = BASE_URL + 'tracker'

export async function registerUser(queryString: string) {
    return await axios.post(URL_REGISTER + queryString)
}

export async function loginUser(json: { [k: string]: string }) {
    return await axios.post(URL_LOGIN, JSON.stringify(json), {
        headers: {
            'Content-Type': 'application/json'
        }
    })
}

// TODO: fix types
export async function getUserDetails(props: any) {
    const [_key, token] = props.queryKey
    const res = await axios.get(URL_USER, {
        headers: { Authorization: `bearer ${token}` }
    })
    if (res.status === 200) {
        store.dispatch(updateUser({ data: res.data, token }))
    }
    return res
}

export async function registerBook(json: { [k: string]: string }) {
    const token = ls.get()
    return await axios.post(`${URL_BOOK}/register`, JSON.stringify(json), {
        headers: {
            'Content-Type': 'application/json',
            Authorization: `bearer ${token}`
        }
    })
}

type GetBookPropTypes = {
    queryKey: string[]
}
export async function getBooks(props: GetBookPropTypes) {
    const token = ls.get()
    const [_key, userId] = props.queryKey
    return await axios.get(`${URL_BOOK}?userId=${userId}`, {
        headers: {
            'Content-Type': 'application/json',
            Authorization: `bearer ${token}`
        }
    })
}

export async function deleteBook(query: string) {
    const token = ls.get()
    return await axios.delete(`${URL_BOOK}/delete${query}`, {
        headers: {
            Authorization: `bearer ${token}`
        }
    })
}

export async function updateBook(json: { id: number; history: string }) {
    const token = ls.get()
    return await axios.patch(`${URL_BOOK}/update`, JSON.stringify(json), {
        headers: {
            'Content-Type': 'application/json',
            Authorization: `bearer ${token}`
        }
    })
}
export async function getWords(props: GetBookPropTypes) {
    const token = ls.get()
    const [_key, userId] = props.queryKey
    return await axios.get(`${URL_TRACKER}/words?userId=${userId}`, {
        headers: {
            'Content-Type': 'application/json',
            Authorization: `bearer ${token}`
        }
    })
}

export async function getSightWords(props: GetBookPropTypes) {
    const token = ls.get()
    const [_key, userId] = props.queryKey
    return await axios.get(`${URL_SIGHT_WORDS}?userId=${userId}`, {
        headers: {
            'Content-Type': 'application/json',
            Authorization: `bearer ${token}`
        }
    })
}

export async function updateTracker(json: {
    userId: number
    libId: string
    bookId: number
    history: HistoryTypes[]
}) {
    console.log('updateTracker:', json)
    const token = ls.get()
    return await axios.patch(`${URL_TRACKER}/update`, JSON.stringify(json), {
        headers: {
            'Content-Type': 'application/json',
            Authorization: `bearer ${token}`
        }
    })
}
