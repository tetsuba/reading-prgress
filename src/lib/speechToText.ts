export default (function speechToText(SPR) {
    if (SPR) {
        const recognition: SpeechRecognition = new SPR()
        recognition.continuous = true
        recognition.lang = 'en-US'
        recognition.interimResults = true // This triggers the onResults a lot faster
        recognition.maxAlternatives = 1

        recognition.onaudiostart = () => console.log('[START]: Audio')
        recognition.onsoundstart = () => console.log('[START]: Sound')
        recognition.onspeechstart = () => console.log('[START]: Speech')
        recognition.onspeechend = () => console.log('[END]: Speech')
        recognition.onsoundend = () => console.log('[END]: Sound')
        recognition.onaudioend = () => console.log('[END]: Audio')

        recognition.onnomatch = () => console.log('[No Match]')
        recognition.onerror = () => console.log('[No Error]')
        recognition.onstart = () => console.log('[Start]')
        recognition.onend = () => console.log('[End]')
        recognition.onresult = () => console.log('[RESULT]')

        return recognition
    }
})(window.SpeechRecognition || webkitSpeechRecognition)
