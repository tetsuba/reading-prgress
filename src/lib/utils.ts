type ValueType = {
    value: string
}

export interface RegistrationFormTypes extends EventTarget {
    firstName?: ValueType
    lastName?: ValueType
    email?: ValueType
    password?: ValueType
}

export function formDataToQueryString(target: RegistrationFormTypes): string {
    const formData: { [key: string]: string } = {
        firstName: target.firstName ? target.firstName.value : '',
        lastName: target.lastName ? target.lastName.value : '',
        email: target.email ? target.email.value : '',
        password: target.password ? target.password.value : ''
    }

    return Object.keys(formData)
        .map((key) => `${key}=${formData[key]}`)
        .join('&')
}

interface LoginFormTypes extends EventTarget {
    email?: ValueType
    password?: ValueType
}
export function formDataToObject(target: LoginFormTypes) {
    return {
        username: target.email?.value,
        password: target.password?.value
    }
}

interface LoginErrorMessageTypes extends Error {
    response?: {
        data: {
            error: string
        }
    }
}

export function getErrorMessage(error: LoginErrorMessageTypes): string {
    if (
        error &&
        error.response &&
        error.response.data &&
        error.response.data.error
    ) {
        return error.response.data.error
    }
    return ''
}

export function delay(time: number) {
    return new Promise((resolve) => setTimeout(() => resolve('success'), time))
}

interface StoryFormTypes extends EventTarget {
    title?: ValueType
    story?: ValueType
}
export function mutateRegisterBookData(
    target: StoryFormTypes,
    userId: number
): { [key: string]: number | string | undefined } {
    return {
        userId,
        title: target.title?.value,
        story: target.story?.value
    }
}
