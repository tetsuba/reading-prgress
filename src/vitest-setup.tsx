import { QueryClient, QueryClientProvider } from 'react-query'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from './store/store'
import { updateUser } from './store/user/userSlice'
import { addBooks } from './store/books/booksSlice'
import { addStudents } from './store/students/studentsSlice'
import { studentsMockData } from '../tests/mockData/students'
import booksMockData from '../tests/mockData/books'
import { userMockData } from '../tests/mockData/user'

const portal = document.createElement('div')
portal.setAttribute('id', 'modal')
document.body.appendChild(portal)

const mockResults = {
    results: [[{ transcript: 'This is a story' }]]
}

class MockSpeech {
    start() {
        // @ts-ignore
        this.onresult(mockResults)
    }
    stop = vi.fn()
    onresult = vi.fn()
    onend = vi.fn()
}

Object.defineProperty(window, 'webkitSpeechRecognition', {
    writable: true,
    value: vi.fn().mockImplementation(() => new MockSpeech())
})

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            staleTime: 0,
            cacheTime: 0,
            retryDelay: 1,
            retry: 0
        }
    }
})

type PropTypes = {
    pathname: string
    children: JSX.Element
    notAuthenticated?: boolean
}

export function updateStoreWithMockData() {
    store.dispatch(updateUser(userMockData))
    store.dispatch(addBooks(booksMockData))
    store.dispatch(addStudents(studentsMockData))
}

export function WrapperWith_Store_Query_Router(props: PropTypes) {
    if (!props.notAuthenticated) updateStoreWithMockData()
    return (
        <Provider store={store}>
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={[props.pathname]}>
                    {props.children}
                </MemoryRouter>
            </QueryClientProvider>
        </Provider>
    )
}

export function WrapperWith_Store_Router(props: PropTypes) {
    return (
        <Provider store={store}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </Provider>
    )
}

export function WrapperWithQueryAndRouter(props: PropTypes) {
    return (
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </QueryClientProvider>
    )
}

export function WrapperWithQuery(props: { children: JSX.Element }) {
    return (
        <QueryClientProvider client={queryClient}>
            {props.children}
        </QueryClientProvider>
    )
}

export function WrapperWithRouter(props: PropTypes) {
    return (
        <MemoryRouter initialEntries={[props.pathname]}>
            {props.children}
        </MemoryRouter>
    )
}
