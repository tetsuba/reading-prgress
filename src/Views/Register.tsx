import { useMutation } from 'react-query'
import { registerUser } from '../lib/service'
import { formDataToQueryString, getErrorMessage } from '../lib/utils'

// COMPONENTS
import Label from '../Components/Form/Label'
import Input from '../Components/Form/Input'
import Button from '../Components/Button/Button'
import ErrorMessage from '../Components/Form/ErrorMessage'
import Header from '../Components/Header/Header'
import Main from '../Components/Main/Main'

export default function Register() {
    const mutation = useMutation(registerUser)

    return (
        <>
            <Header text="Register" />
            <Main>
                <div className="mx-auto max-w-3xl rounded-lg border-dashed border-gray-200 sm:border-4">
                    <div className="sm:p-6">
                        {mutation.isLoading && (
                            <p data-testid="loading-user">
                                Registering user...
                            </p>
                        )}
                        {mutation.isSuccess && (
                            <div data-testid="success-message">
                                <p>Registration Completed.</p>
                                <p>Please click on the log in button...</p>
                            </div>
                        )}

                        {!mutation.isSuccess && (
                            <form
                                data-testid="register-form"
                                className="space-y-6"
                                onSubmit={(event) => {
                                    event.preventDefault()
                                    const queryString = formDataToQueryString(
                                        event.target
                                    )
                                    mutation.mutate(queryString)
                                }}
                            >
                                <div>
                                    <Label htmlFor="firstName">
                                        First Name:
                                    </Label>
                                    <Input
                                        template="text"
                                        type="text"
                                        name="firstName"
                                        placeholder="first name"
                                    />
                                </div>
                                <div>
                                    <Label htmlFor="lastName">Last Name:</Label>
                                    <Input
                                        template="text"
                                        type="text"
                                        name="lastName"
                                        placeholder="last name"
                                    />
                                </div>
                                <div>
                                    <Label htmlFor="email">Your email:</Label>
                                    <Input
                                        template="text"
                                        type="email"
                                        name="email"
                                        placeholder="name@company.com"
                                    />
                                </div>
                                <div>
                                    <Label htmlFor="password">
                                        Your password:
                                    </Label>
                                    <Input
                                        template="text"
                                        type="password"
                                        name="password"
                                        placeholder="••••••••"
                                    />
                                </div>

                                <ErrorMessage
                                    show={mutation.isError}
                                    className="mt-1 block"
                                >
                                    {getErrorMessage(mutation.error as Error)}
                                </ErrorMessage>

                                <div className="flex sm:justify-end">
                                    <Button
                                        className="w-full sm:w-fit"
                                        template="primary"
                                        type="submit"
                                    >
                                        Register
                                    </Button>
                                </div>
                            </form>
                        )}
                    </div>
                </div>
            </Main>
        </>
    )
}
