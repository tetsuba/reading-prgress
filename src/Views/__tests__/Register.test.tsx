import { act, fireEvent, render, screen, waitFor } from '@testing-library/react'
import Register from '../Register'
import { WrapperWithQuery } from '../../vitest-setup'
import axios from 'axios'
import { delay } from '../../lib/utils'
vi.mock('axios')

describe('Register', () => {
  test('should render', () => {
    const { asFragment } = render(
      <WrapperWithQuery>
        <Register />
      </WrapperWithQuery>
    )
    expect(asFragment()).toMatchSnapshot()
  })
  test('should return success', async () => {
    // @ts-ignore
    axios.post.mockResolvedValue(delay(10))

    const eventTarget = {
      target: {
        firstName: { value: 'bob' },
        lastName: { value: 'bob' },
        password: { value: '1234' },
        email: { value: 'a@a.com' },
      },
    }
    render(
      <WrapperWithQuery>
        <Register />
      </WrapperWithQuery>
    )

    act(() => {
      fireEvent.submit(screen.getByTestId('register-form'), eventTarget)
    })

    await expect(screen.findByTestId('loading-user'))
    await waitFor(() => expect(screen.getByTestId('success-message')))
  })
  test('should return an error', async () => {
    // @ts-ignore
    axios.post.mockRejectedValueOnce({ error: { message: 'error' } })

    const eventTarget = {
      target: {
        firstName: { value: '' },
        lastName: { value: '' },
        password: { value: '' },
        email: { value: '' },
      },
    }
    render(
      <WrapperWithQuery>
        <Register />
      </WrapperWithQuery>
    )

    act(() => {
      fireEvent.submit(screen.getByTestId('register-form'), eventTarget)
    })

    await waitFor(() => expect(screen.getByTestId('error-message')))
  })
})
