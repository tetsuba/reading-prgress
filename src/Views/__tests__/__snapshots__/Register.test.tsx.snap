// Vitest Snapshot v1

exports[`Register > should render 1`] = `
<DocumentFragment>
  <header
    class="bg-white shadow"
  >
    <div
      class="mx-auto flex max-w-7xl justify-between py-6 px-4 sm:px-6 lg:px-8"
    >
      <h1
        class="text-3xl font-bold tracking-tight text-gray-900"
      >
        Register
      </h1>
    </div>
  </header>
  <main>
    <div
      class="mx-auto max-w-7xl py-6 md:px-6 lg:px-8"
    >
      <div
        class="mx-auto max-w-3xl rounded-lg border-dashed border-gray-200 sm:border-4"
      >
        <div
          class="sm:p-6"
        >
          <form
            class="space-y-6"
            data-testid="register-form"
          >
            <div>
              <label
                class="text-sm font-medium text-gray-900 dark:text-white "
                for="firstName"
              >
                First Name:
              </label>
              <input
                class="bg-gray-50 border border-gray-300  text-gray-900 text-sm rounded-lg block w-full p-2.5 focus:ring-blue-500 focus:border-blue-500 "
                name="firstName"
                placeholder="first name"
                template="text"
                type="text"
              />
            </div>
            <div>
              <label
                class="text-sm font-medium text-gray-900 dark:text-white "
                for="lastName"
              >
                Last Name:
              </label>
              <input
                class="bg-gray-50 border border-gray-300  text-gray-900 text-sm rounded-lg block w-full p-2.5 focus:ring-blue-500 focus:border-blue-500 "
                name="lastName"
                placeholder="last name"
                template="text"
                type="text"
              />
            </div>
            <div>
              <label
                class="text-sm font-medium text-gray-900 dark:text-white "
                for="email"
              >
                Your email:
              </label>
              <input
                class="bg-gray-50 border border-gray-300  text-gray-900 text-sm rounded-lg block w-full p-2.5 focus:ring-blue-500 focus:border-blue-500 "
                name="email"
                placeholder="name@company.com"
                template="text"
                type="email"
              />
            </div>
            <div>
              <label
                class="text-sm font-medium text-gray-900 dark:text-white "
                for="password"
              >
                Your password:
              </label>
              <input
                class="bg-gray-50 border border-gray-300  text-gray-900 text-sm rounded-lg block w-full p-2.5 focus:ring-blue-500 focus:border-blue-500 "
                name="password"
                placeholder="••••••••"
                template="text"
                type="password"
              />
            </div>
            <div
              class="flex sm:justify-end"
            >
              <button
                class="rounded-lg font-medium text-base text-center px-5 py-2.5 text-white bg-blue-700 hover:bg-blue-800 w-full sm:w-fit"
                template="primary"
                type="submit"
              >
                Register
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
